ТЗ-14

IIExpanse opened the pull request:
Андрей, добрый день.
Отправляю на проверку ТЗ 14 спринта. По требованиям все реализовано, внутренние тесты реализованы частично, оставшиеся допишу в рамках следующего задания - на проект уже ушло очень много времени и следующее ТЗ, насколько я понимаю, во многом посвящено покрытию всего реализованного функционала тестами.
Для работы подключалась PostgreSql, поскольку акцент в ТЗ был на использовании ее драйвера.

IIExpanse committed:
feat: Добавлены функции бронирования, оставления отзывов, настроена работа с БД через JPA-репозитории.

IIExpanse committed:
style: мелкие изменения стиля по требованию тестов

IIExpanse committed:
style: bless style auto-checkers..

IIExpanse committed:
fix: устранена ошибка в маппере после стилевых правок

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,160 @@
+    private final BookingService bookingService;
+    private final UserService userService;
+    private final ItemService itemService;
+    private final BookingMapper mapper;
```
маппер стоит перенести на уровень сервисов, контроллер должен вызывать только методы сервисов без дополнительной логики

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,160 @@
+    @PostMapping
+    public ResponseEntity<BookingDto> addBooking(@RequestHeader(name = "X-Sharer-User-Id") long bookerId,
+                                                 @RequestBody @Valid BookingDtoRequest bookingDto) {
+        LocalDateTime start = bookingDto.getStart();
```
вся эта логика должна быть в сервисе

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,160 @@
+        return ResponseEntity.ok(mapper.mapToDto(booking, bookingService.determineStatus(booking)));
+    }
+
+    private BookingStatus parseStatus(String state) {
```
приватные методы также только в сервисах

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -0,0 +1,29 @@
+import ru.practicum.shareit.user.model.User;
+
+@Mapper(componentModel = "spring")
+public interface BookingMapper {
```
отлично, большой плюс за использование мапперов

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,68 @@
+            "AND (b.approved = ?3 OR ?3 IS NULL AND b.approved IS NULL)" +
+            "ORDER BY b.startTime DESC"
+    )
+    Collection<Booking> getAllByBookerIdOrItemOwnerIdAndApprovedIsOrderByStartTimeDesc(
```
большой плюс за JPQL-запросы. По идее, если мы используем такой запрос, то название самого метода может быть любым, т.е. его можно сделать покороче

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/service/impl/BookingServiceImpl.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Override
+    public Booking addBooking(Booking booking) {
```
возвращать в контроллер лучше всего именно dto, т.к. возврат объекта модели целиком является плохой практикой

andrew-larkin commented on src/main/java/ru/practicum/shareit/booking/service/BookingStatus.java:
```diff
@@ -0,0 +1,11 @@
+package ru.practicum.shareit.booking.service;
+
+public enum BookingStatus {
```
этот enum лучше перенести в папку с моделями

andrew-larkin reviewed (changes requested):
Привет! Отлично, очень качественная работа. Нам остается немного привести ее в порядок, я все оставил комментарии по коду только в папке booking, но они актуальны для всех остальных папок. Небольшой рефакторинг и мы будем готовы двигаться дальше.

IIExpanse committed:
fix: 1. Из контроллеров убрана вся логика, кроме вызовов сервисов и формирования ответом.
2. Enum BookingStatus перемещен в пакет с DTO-объектами.
3. Часть методов с длинными названиями, использующих JPQL, переименована.

Nagash40 commented:
Из контроллеров вынесено все, кроме обычных вызовов сервисов и отправки ответов, все затронутые рефакторингом тесты изменены для работы с новой логикой.
Из небольших изменений переименована часть длинных методов, работающих с JPQL,, и перемещен в директорию с DTO enum со статусами.
